### ISSUE TITLE
Improve compile_commands refresh time.

### ISSUE THREAD

**Issue by rshanor** [hedronvision/bazel-compile-commands-extractor#24]
Hi,

First thanks for an awesome tool! Super easy to set up and use. I am new to clangd and bazel, so sorry in advance for any dumb questions. 

Do you have any ideas about how to cache results and speed up regeneration time? I work across a large monorepo and even after specifying some broad targets in `refresh_compile_commands`, the refresh command takes about 9min. 

Happy to help implement a fix if you have suggestions. Thanks for the advice!

**Comment by cpsauer** [comment]
Hey, Rick! Thanks for playing around with this tool--and for your kind words. Glad to hear you're finding it easy to get started with. And I appreciate your willingness to help. [Sounds like quite a big repo! You've got me curious about what it is.]

I definitely have ideas, and this is something we've thought a fair amount about :) Maybe let's peel back the onion a little bit to see ways we could get things faster for you? 

### What's the runtime going towards? How should it really be fixed?

At the root, most of the runtime of this tool is working around https://github.com/clangd/clangd/issues/123. If you're looking to really dive in (or make the case for the need), that's really the best place to fix things properly.

What's going on is probably best explained in #5. (Github: Duplicate of #5) But basically without this workaround, clangd doesn't proactively figure out how headers are compiled, unless you've already got the corresponding implementation file open. This leads to weirdness in a bunch of cases.

...so this tool proactively runs the preprocessor (in parallel) over every source file we find, issuing entries for headers, so clangd knows how to understand them when they're open. This prevents the clangd issue from propagating up to bug you, the user, but it is responsible for most of the runtime. If you're looking to go deeper, there are lots of notes about this near the top of [extract.py](https://github.com/hedronvision/bazel-compile-commands-extractor/blob/main/extract.py).

So you're getting a fair amount of bang for the buck on that runtime. I promise :) And human time is a lot more expensive than CPU time.

But it's just a workaround, doing work up front that clangd should be doing and caching in the background. Really, we eventually want clangd to be the one understanding the include graph. We extract commands from Bazel; clangd does the analysis, running it as part of its background indexing jobs, and including it in its caches. Bada bing, bada boom.

### Other interim fixes:

If engaging with https://github.com/clangd/clangd/issues/123 seems too messy for right now (fair enough), there are some other possible fixes.

The first and simplest (as in #5), is that you might find that you don't have to refresh compile_commands.json very often at all, making the current runtime okay. If you're adding files, clangd should make pretty decent guesses at completions, using commands from nearby files. And if you're deleting files, there's not a problem. So you may not need to rerun Refresh.sh on every change to BUILD files. Maybe this is something you run every so often when you can spare the time. Basically just using compile_commands.json as the cache :)

The second is to index only the part of the codebase you're working on. Sounds like you've already tried that with `refresh_compile_commands`.

Third and finally, we could improve our workaround to https://github.com/clangd/clangd/issues/123, but I don't love the remaining options there. As above, some notes on that in [extract.py](https://github.com/hedronvision/bazel-compile-commands-extractor/blob/main/extract.py). Some other sub-options for improving the workaround: We could add caching in the build tree, to avoid doing this every time. (But kinda bad form, bazel-wise, and it's very nice for simplicity that this tool is a pure function hitting bazel's action cache. Once upon a time, we actually got bazel to cache some of this for us, with an older, extra_action-based approach--see this part of the [ImplementationReadme](https://github.com/hedronvision/bazel-compile-commands-extractor/blob/main/ImplementationReadme.md#choice-of-strategy-for-listening-to-bazel)--but it was still way slower than this approach, and had a number of other drawbacks.) Or, as another not-great sub-option, we could add an attribute to turn off the header finding (I know @lummax was thinking about doing this), but then you won't benefit from the workaround. If you're looking to quickly get a sense for just how much of the runtime is going into this workaround and what problems the clangd issue causes, you could quickly experiment by commenting out the header finding parts--just to try. But yeah, the further speed improvements to the workaround that I can see aren't *great*, which is why I haven't done them. If you see an opportunity that I've missed, please say something!

### Summary

Anyway, to sum: Sorry the runtime is being a bother. I'd eliminate it if I could. But I think the workaround is well worth its runtime; you're getting a fair amount for the work, which is why it's being done automatically for you. A good fix would really be in clangd for https://github.com/clangd/clangd/issues/123, eliminating our slow workaround. Otherwise, maybe one of the other options above would work for you? Let me know what you think!

**Comment by cpsauer** [comment]
Duplicate of #5

**Comment by rshanor** [comment]
Wow, thanks for all the info Christopher. Once I start using the tool more in my day to day workflow, I will probably have more concrete questions and ideas. Right now I dont have a great sense of how often I need to refresh this index in order to get the performance I want. Another option could always be to periodically generate this file via CI so that devs can pull as needed. In the past I have used the Microsoft vscode c++ plugin, which never seems to perform as well as I want it too. Hence why I am trying something new.

**Comment by cpsauer** [comment]
Sounds good! Please keep me posted.

(Quick notes back: As above, I suspect refresh only occasionally--but that you'll want it good when you do refresh. Same experience with the Microsoft one last I used it--it was choking on a number of common non-windows flags in the compile commands. The CI idea is a creative one!)

----

I'm going to close this for now to keep #5 the main issue for this (also linking back here). But please do reply back with what you find and how things go. I'll open it right back up if you need me or if we get going on a fix.

**Comment by cpsauer** [comment]
@rshanor, I just thought of a good, fairly easy way to cache things and speed them up. See https://github.com/hedronvision/bazel-compile-commands-extractor/issues/5#issuecomment-1091008003

[I'm also curious what you ended up learning from continued use!]

**Comment by cpsauer** [comment]
@rshanor, heads that things should be way faster now. 

I just landed some code that caches header file finding and for clang/gcc, piggybacks on Bazel's dependency cache. There are also some new options to ignore external workspaces. See the README for those

Anyway, that should be a huge improvement in speed. Hoping that together those might make things significantly better and more usable for you!

**Comment by kkpattern** [comment]
Maybe we can add a flag for turning `_get_headers` off? I think some tools do not need headers to be included in compilation database. Like clang-tidy, include-what-you-use. The cache helped a lot for clang toolchain. But on Windows it's just too slow for our relatively large project.

By the way, thanks very much for this awesome tool! We succeeded to generate compilation database on our first try! For two cross-compile target platforms! Very solid tool!

**Comment by kkpattern** [comment]
Maybe we can use aspect to extract all the header files from a target before we run aquery for the same target?

**Comment by cpsauer** [comment]
Hey @kkpattern! You're very welcome. Glad it's working for you guys :)

Totally agree re allowing people to turn off header extraction. Would `exclude_headers = "all"` work for you? (See [here in the README](https://github.com/hedronvision/bazel-compile-commands-extractor#4-using-ccls-or-another-tool-that-unlike-clangd-doesnt-want-or-need-headers-in-compile_commandsjson) and slightly below about speeding things up.)

For MSVC, we'll still cache between runs, so only the first run will be slow. Any chance fast incremental is good enough for you guys?
[We can't hit bazel's cache, I think, because Bazel doesn't store header information on disk with MSVC like they do on other platforms. Though if you find it cached somewhere in bazel-out, please do let me know!]

**Comment by cpsauer** [comment]
[And my past experience with trying to get headers from aquery and aspects is that you end up getting a pretty incomplete set.]

**Comment by kkpattern** [comment]
> Would `exclude_headers = "all"` work for you? (See [here in the README](https://github.com/hedronvision/bazel-compile-commands-extractor#4-using-ccls-or-another-tool-that-unlike-clangd-doesnt-want-or-need-headers-in-compile_commandsjson) and slightly below about speeding things up.)

Oh somehow I missed the `exclude_headers = "all"` doc. Works like a charm for us. Thanks!
